core/src/main/java/io/grpc/internal/AbstractStream2.java

```diff
@@ -0,0 +1,261 @@
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import io.grpc.Codec;
+import io.grpc.Compressor;
+import io.grpc.Decompressor;
+
+import java.io.InputStream;
+
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Only called from application thread.
+ */
+public abstract class AbstractStream2 implements Stream {
+  protected abstract MessageFramer framer();
+
+  protected abstract TransportState transportState();
+
+  @Override
+  public final void setMessageCompression(boolean enable) {
+    framer().setMessageCompression(enable);
+  }
+
+  @Override
+  public final void writeMessage(InputStream message) {
+    checkNotNull(message);
+    if (!framer().isClosed()) {
+      framer().writePayload(message);
+    }
+  }
+
+  @Override
+  public final void flush() {
+    if (!framer().isClosed()) {
+      framer().flush();
+    }
+  }
+
+  /**
+   * Closes the underlying framer.
+   */
+  protected final void endOfMessages() {
+    framer().close();
+  }
+
+  @Override
+  public final void setCompressor(Compressor compressor) {
+    framer().setCompressor(checkNotNull(compressor, "compressor"));
+  }
+
+  @Override
+  public final void setDecompressor(Decompressor decompressor) {
+    transportState().setDecompressor(decompressor);
+  }
+
+  @Override
+  public final boolean isReady() {
+    if (framer().isClosed()) {
+      return false;
+    }
+    return transportState().isReady();
+  }
+
+  /**
+   * Event handler to be called by the subclass when a number of bytes are being queued for sending
+   * to the remote endpoint.
+   *
+   * @param numBytes the number of bytes being sent.
+   */
+  protected final void onSendingBytes(int numBytes) {
+    transportState().onSendingBytes(numBytes);
+  }
+
+  /**
+   * Only called from transport thread (except for private interactions with AbstractStream2).
+   */
+  public abstract static class TransportState implements MessageDeframer.Listener {
```

Comment:
The name of this class is more properly AbstractStream2.TransportState. When said that way it seems fairly obvious this is stream-specific state. I understand there are times it is necessary to stutter in class names, but I tend to limit to that to cases when the class name will commonly cause a name clash with another class (I'm looking at you `java.awt.List`!).